pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Schema.java

```diff
@@ -167,25 +168,47 @@ default T decode(byte[] bytes, byte[] schemaVersion) {
     }
 
     /**
-     * Create a Avro schema type by extracting the fields of the specified class.
+     * Create a  Avro schema type by default configuration of the class
      *
      * @param clazz the POJO class to be used to extract the Avro schema
      * @return a Schema instance
      */
     static <T> Schema<T> AVRO(Class<T> clazz) {
-        return DefaultImplementation.newAvroSchema(clazz);
+        return DefaultImplementation.newAvroSchema(new SchemaDefinition<>(clazz));
     }
 
     /**
-     * Create a JSON schema type by extracting the fields of the specified class.
+     * Create a Avro schema type with schema definition
+     *
+     * @param schemaDefinition the definition of the schema
+     * @return a Schema instance
+     */
+    static <T> Schema<T> AVRO(SchemaDefinition<T> schemaDefinition) {
+        return DefaultImplementation.newAvroSchema(schemaDefinition);
+    }
+
+
+    /**
+     * Create a JSON schema type by default configuration of the class
      *
      * @param clazz the POJO class to be used to extract the JSON schema
      * @return a Schema instance
      */
     static <T> Schema<T> JSON(Class<T> clazz) {
-        return DefaultImplementation.newJSONSchema(clazz);
+        return DefaultImplementation.newJSONSchema(new SchemaDefinition<>(clazz));
+    }
+
+    /**
+     * Create a JSON schema type with schema definition
+     *
+     * @param schemaDefinition the definition of the schema
+     * @return a Schema instance
+     */
+    static <T> Schema<T> JSON(SchemaDefinition schemaDefinition) {
```

Comment:
I think that JSON is a bit different compared with AVRO in that it does not have a canonical schema representation. In AVRO, the schema def is the source of truth, but for JSON the pojo is typically the source of truth (and people use annotations to customize the specific fields). 

For now I'd prefer to not add it. We can always add it later if there are good reasons for it (but we'd not be able to take it out once it's in there).

Reply:
AllowNull probably doesn't make a lot sense to JSON. but having the ability to create a schema definition is good for query engines using pulsar. so I would prefer keeping it for consistency as AVRO, as it would allow people to customize how schema is generated.

Reply:
A part from having this difference in the allow null, I think there is a more fundamental difference between JSON and AVRO.

In JSON, the AVRO schema is extracted and used for the validation, but the JSON serialization is done separately. In this case the POJO and the serialized data could be different from the Avro schema that was validated with broker.

That's not a problem with Avro since the schema object is used at serialization time so it doesn't matter which POJO we're passing there that the result will be correct (or user gets exception). 

For this I'm suggesting to avoid adding the option for JSON right now and defer it until there's a clear use case for it (adding is always easy).

My guess is that this will probably not be needed as users will not already have a precise Avro schema representation for the data used in other systems. Rather, it's common practice in most languages to define POJOs for the JSON data.

Reply:
Discussed with @merlimat offline about this. Both of us agree on having the same SchemaDefinition for both JSON and Avro, but make clear what are the use cases for these methods. 

The SchemaDefinition (e.g. AllowNull or schema string in #3766) for JSON probably looks not useful initially. But this is useful for an integration between Pulsar and other data systems with their own schema systems (e.g. Hive). The schema is useful defined and generated by `GenericSchemaBuilder` and converting the metadata definition from other system to pulsar's schema system. The generated schema will be used as `source-of-truth` for being consistent across pulsar and other data systems. So we need a mechanism for this generated schema to be attached to JSON schema when it is used in a POJO form. Otherwise if we don't do so, the schema parsed from POJO might not exactly have the constraints that we converted from other schema systems. And it will cause incompatible schema exceptions.

With that being said, we will keep SchemaDefinition for Schema.JSON in this PR, so it can be used in #3766. But we need to change `SchemaDefinition` to a builder pattern to allow extensibility. /cc @congbobo184 

@merlimat please correct me if I missed anything from our conversation.